### PR TITLE
Derive pipeline state from workflows

### PR DESCRIPTION
**Pull request by polymeris**
~~This is based on top of #38, so just look at the top commit https://github.com/CircleCI-Public/gitlab-single-org-connector/pull/40/commits/f3c319e7f1e6453c1141f1142af6c7993a098610~~

### CODE REVIEW

**Comment by conormcd** [comment]
Having thought about this a bit (and I'm sorry for not thinking about this before!) we should probably make the `PipelineStatusPoller` a `WorkflowStatusPoller` and track/update each workflow separately.

Commits can have multiple statuses posted back to them (and this is what we do for GitHub) so it makes sense to do them independently.

**Comment by polymeris** [comment]
> Having thought about this a bit (and I'm sorry for not thinking about this before!) we should probably make the `PipelineStatusPoller` a `WorkflowStatusPoller` and track/update each workflow separately.

Makes sense to me.

> Commits can have multiple statuses posted back to them (and this is what we do for GitHub) so it makes sense to do them independently.

I am not sure we can post multiple statuses to GitLab? https://docs.gitlab.com/ee/api/commits.html#post-the-build-status-to-a-commit
How are there differentiated?

**Comment by polymeris** [comment]
Rebased on master

File: src/main/java/com/circleci/connector/gitlab/singleorg/client/CircleCi.java [hunk 1]
```diff
@@ -85,15 +95,40 @@ public Pipeline refreshPipeline(Pipeline pipeline) {
...
+      return State.RUNNING;
+    }
+    if (workflowStates.contains(State.PENDING)) {
+      return State.PENDING;
     }
+    if (workflowStates.contains(State.SUCCESS)) {
+      return State.SUCCESS;
+    }
+    return State.PENDING;
+  }
```
Comment (gordonsyme) [inline src/main/java/com/circleci/connector/gitlab/singleorg/client/CircleCi.java]: These aren't pipeline states. Pipelines don't aggregate the state of their constituent workflows. What's the issue being solved?

Comment (polymeris) [reply src/main/java/com/circleci/connector/gitlab/singleorg/client/CircleCi.java]: > Pipelines don't aggregate the state of their constituent workflows. 
They do in this PR, but we are discussing changing that if the GitLab API allows reporting of individual pipelines.

**Comment by polymeris** [comment]
> > I am not sure we can post multiple statuses to GitLab? https://docs.gitlab.com/ee/api/commits.html#post-the-build-status-to-a-commit
> > How are there differentiated?
> 
> The `name` or `context` field in the POST. We should use the workflow name for that.

Ok. I will work on a new PR today.

File: src/main/java/com/circleci/connector/gitlab/singleorg/client/CircleCi.java (hunk 1 above)
Comment (gordonsyme) [reply src/main/java/com/circleci/connector/gitlab/singleorg/client/CircleCi.java]: Right, I'm saying this PR is doing the wrong thing :)

Comment (polymeris) [reply src/main/java/com/circleci/connector/gitlab/singleorg/client/CircleCi.java]: Yes, I'll experiment with a different approach today. Adding the concept of a workflow to the connector and passing the individual statuses along to GitLab.

**Comment by polymeris** [comment]
PR #41 replaces this

File: src/main/java/com/circleci/connector/gitlab/singleorg/client/CircleCi.java (hunk 1 above)
Comment (polymeris) [reply src/main/java/com/circleci/connector/gitlab/singleorg/client/CircleCi.java]: Have a look at PR #41 if you are interested, @gordonsyme

**Pull request closed without merging**